udf/mrn_udf_normalize.cpp

```diff
@@ -0,0 +1,211 @@
+    const char *action;
+    if (current_db_path) {
+      action = "open database";
+      mrn::Database *db;
+      int error = mrn_db_manager->open(current_db_path, &db);
+      if (error == 0) {
+        info->db = db->get();
+        grn_ctx_use(info->ctx, info->db);
+        info->use_shared_db = true;
+      }
+    } else {
+      action = "create anonymous database";
+      info->db = grn_db_create(info->ctx, NULL, NULL);
+      info->use_shared_db = false;
+    }
+    if (!info->db) {
+      sprintf(message,
+              "mroonga_normalize(): failed to %s: %s",
+              action,
+              info->ctx->errbuf);
+      goto error;
+    }
+  }
+
+  if (args->arg_count == 1) {
+    info->normalizer = grn_ctx_get(info->ctx, DEFAULT_NORMALIZER_NAME, -1);
+  } else {
+    info->normalizer = grn_ctx_get(info->ctx, args->args[1], args->lengths[1]);
+  }
+  if (!info->normalizer) {
+    sprintf(message, "mroonga_normalize(): nonexistent normalizer %.*s",
+            (int)args->lengths[1], args->args[1]);
+    goto error;
+  }
+  info->flags = 0;
+
+  result_str = &(info->result_str);
+  mrn::encoding::set_raw(info->ctx, system_charset_info);
+  result_str->set_charset(system_charset_info);
+
+  initid->ptr = (char *)info;
+
+  return FALSE;
+
+error:
+  if (info) {
+    if (!info->use_shared_db) {
+      grn_obj_close(info->ctx, info->db);
+    }
+    mrn_context_pool->release(info->ctx);
+    my_free(info);
+  }
+  return TRUE;
+}
+
+MRN_API char *mroonga_normalize(UDF_INIT *initid, UDF_ARGS *args, char *result,
+                                unsigned long *length, char *is_null, char *error)
+{
+  st_mrn_normalize_info *info = (st_mrn_normalize_info *)initid->ptr;
+  grn_ctx *ctx = info->ctx;
+  String *result_str = &(info->result_str);
+  
+  if (!args->args[0]) {
+    *is_null = 1;
+    return NULL;
+  }
+
+  result_str->length(0);
+  {
+    char *target = args->args[0];
+    unsigned int target_length = args->lengths[0];
+    grn_obj *grn_string;
+    const char *normalized;
+    unsigned int normalized_length_in_bytes;
+    unsigned int normalized_n_characters;
+
+    grn_string = grn_string_open(ctx,
+                                 target, target_length,
+                                 info->normalizer, info->flags);
+    grn_string_get_normalized(ctx, grn_string,
+                              &normalized,
+                              &normalized_length_in_bytes,
+                              &normalized_n_characters);
+    if (result_str->reserve(normalized_length_in_bytes)) {
+      my_error(ER_OUT_OF_RESOURCES, MYF(0), HA_ERR_OUT_OF_MEM);
+      goto error;
+    }
+    result_str->q_append(normalized, normalized_length_in_bytes);
+    result_str->length(normalized_length_in_bytes);
+    grn_obj_unlink(ctx, grn_string);
+  }
+  *is_null = 0;
+
+  if (ctx->rc) {
+    my_message(ER_ERROR_ON_WRITE, ctx->errbuf, MYF(0));
+    goto error;
+  }
+
+  *length = result_str->length();
+  DBUG_RETURN((char *)result_str->ptr());
```

Comment:
We need more works in `CMakeLists.txt` and `udf/Makefile.am` to use `DBUG_RETURN()`. (Some build flags should be added.)

But the works aren't the work in this pull request. In other words, you should use `return` instead of `DBUG_RETURN()` in this pull request.